tests/utilities/test_grads.py

```diff
@@ -0,0 +1,77 @@
+# Copyright The PyTorch Lightning team.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+from math import isclose, sqrt
+
+import pytest
+import torch
+import torch.nn as nn
+
+from pytorch_lightning.utilities import grad_norm
+
+
+@pytest.mark.parametrize(
+    "norm_type,expected",
+    [
+        # bug for L0 "norm"! total norm should be 5 here
```

Comment:
Why do you say it is a bug?

```python
np.linalg.norm(np.array([3., 2.]), 0)
Out[31]: 2.0
```

which matches the definition of vector 0-norm: `sum(x != 0)`

Reply:
yes, 100% a bug in our implementation, I am sure. Agree your example is fine and I'm not saying `torch.norm` computes an incorrect result. It's our implementation that takes a shortcut for efficiency to avoid recomputing values when computing the total  norm of all parameters combined. That shortcut holds up for p>0 because of properties of the Lp-norm. But since for p=0 it is not a true norm, that formula breaks down. Here is what we do:

Let v0 and v1 be two vectors. 

Individual norms:
```
n0 = norm(v0, p) 
n1 = norm(v1, p)
```

Total norm of all elements combined:
```
t = norm([v0, v1], p)
```

Instead of the above, our implementation does this using the already computed values n0 and n1:
```
t = norm([n0, n1], p)
```

One can show that
```
norm([v0, v1], p) = norm([n0, n1], p) for p > 0
```

But for p = 0 this does not hold true. 
Example:

```
v0 = [-1, 2]
v1 = [3, 4]

```
We will get `n0 = n1 = 2`
but
```
2 = norm([2, 2], 0) != norm([-1, 2, 3, 4], 0) = 4
```

Reply:
This hasn't been an issue because:

https://github.com/PyTorchLightning/pytorch-lightning/blob/44aed17affc396a82ea3bb93345c1892ebaee82b/pytorch_lightning/loops/optimization/optimizer_loop.py#L485

but if we don't want to support 0-norm, it should raise an exception instead of silently skipping it